docs(accessibility-development): Updates documentation as part of content audit. (#3780)
patternfly/patternfly-org: packages/documentation-site/patternfly-docs/content/developer-resources/accessibility-development.md

    - Add a label for the icon in tooltip text that displays on hover and focus.
    - For interactive elements (like `<a>` and `<button>`), where an icon is used as the label instead of text, set a label as an `aria-label`. 

      Example: 

        ```html noLive
        <button class="..." aria-label="Close Dialog">
          <i class="..." aria-hidden="true"></i>
        </button>
        ```

##### **Non-interactive icons**

For **non-interactive icons**, include `.pf-v5-screen-reader` text near the icon. Depending on the component, this text might not be a direct sibling to the icon element.  

#### Images

All images should have alt text so that assistive technology can provide an image description. When an image contains words that are important to understanding the content, the alt text should include those words. This allows the alt text to communicate the same message as the image itself. 

**Note**: Alt text does not necessarily describe the visual characteristics of the image itself. It just needs to convey the same meaning as the image. 

> The exception to this practice is when images are primarily for presentation purposes and are not essential pieces of content. To signify that an image should be skipped by a screen reader, set the value of the alt attribute to an empty string: `alt=””`.

Author: ```suggestion
The exception to this practice is when images are primarily for presentation purposes and are not essential pieces of content. To signify that an image should be skipped by a screen reader, set the value of the alt attribute to an empty string: `alt=””`.

For example, in the alert component, the icon label text is adjacent to the message. This way, when `role="alert"` is added to `.pf-v5-c-alert__body` for dynamically displayed alerts, the type of message is announced along with the message text.
```

section: accessibility
---
import { AccessibilityDatalist } from './accessibility-datalist.js';

PatternFly provides accessible components, but **we can't guarantee that your products will be accessible**. In order to ensure that your product is accessible, you will need to take additional steps during development. For guidance on this development work, you can refer to these guidelines.

## Prioritizing accessibility during development

As you develop your product, it can be helpful to keep accessibility top of mind, by asking yourself a few questions: 

- **Can all users discover and perceive elements?**: If you can see or click on it, then all users should be able to locate and navigate to it. This should include those that use a keyboard or other assistive technology, such as a screen reader.

- **Can all users interact with elements?**. It should be easy for users to use an element once it's in focus. They should be able to interact with elements and initiate actions via keyboard (like pressing **Enter** to initiate a button action) and screen readers.

- **Can all users understand elements?**: It should be clear what actions elements can perform. 

>   - For example, buttons should have visible text that would be clear out of context of the page. If not, it should have an aria-label or accessible name.

Reviewer: Nit not a blocker: could these be combined into a single bullet, similar to the other bullets?

Author: ```suggestion
- **Can all users understand elements?**: It should be clear what actions elements can perform. For example, buttons should have visible text that would be clear out of context of the page. If not, it should have an aria-label or accessible name.
```

## Development techniques

> There are items outside the scope of PatternFly that you should prioritize to ensure accessibility: 

Reviewer: ```suggestion
There are some of the items outside the scope of PatternFly that you should prioritize to ensure accessibility: 
```

Might be me being too cautious with ensuring we don't inadvertently convey "this is all you need to do and nothing else" 😆

> There are items outside the scope of PatternFly that you should prioritize to ensure accessibility: 

Author: ```suggestion
These are some of the items outside the scope of PatternFly that you should prioritize to ensure accessibility: 
```

---
id: Develop for accessibility
section: accessibility
---
import { AccessibilityDatalist } from './accessibility-datalist.js';

PatternFly provides accessible components, but **we can't guarantee that your products will be accessible**. In order to ensure that your product is accessible, you will need to take additional steps during development. For guidance on this development work, you can refer to these guidelines.

## Prioritizing accessibility during development

As you develop your product, it can be helpful to keep accessibility top of mind, by asking yourself a few questions: 

- **Can all users discover and perceive elements?**: If you can see or click on it, then all users should be able to locate and navigate to it. This should include those that use a keyboard or other assistive technology, such as a screen reader.

> - **Can all users interact with elements?**. It should be easy for users to use an element once it's in focus. They should be able to interact with elements and initiate actions via keyboard (like pressing **Enter** to initiate a button action) and screen readers.

Reviewer 2: Not all that important, but I noticed that this line has a period instead of a colon like the others. :)

> - **Can all users interact with elements?**. It should be easy for users to use an element once it's in focus. They should be able to interact with elements and initiate actions via keyboard (like pressing **Enter** to initiate a button action) and screen readers.

Reviewer: ```suggestion
- **Can all users interact with elements?**. It should be easy for users to use an element once it's in focus. They should be able to interact with elements and initiate actions via keyboard (like pressing <kbd>Enter</kbd> to initiate a button action) and screen readers.
```

Author: Ahh I was working locally yesterday and didn't end up pushing, so that's probably what you saw! I just pushed though & I think both your syntax comments should be addressed 😀 @thatblindgeye

> - **Can all users understand elements?**: It should be clear what actions elements can perform. For example, buttons should have visible text that would be clear out of context of the page. If not, it should have an aria-label or accessible name.

Reviewer: ```suggestion
- **Can all users understand elements?**: It should be clear what actions elements can perform. For example, buttons should have visible text that would be clear out of context of the page. If not, it should have an `aria-label` or accessible name.
```

Reviewer: Similar to above, backticks aren't shown in the file around aria-label

---

> id: Develop accessibly

Author: ```suggestion
id: Develop for accessibility
```

> id: Accessible development

Author: ```suggestion
id: Develop accessibly
```

section: accessibility
---
import { AccessibilityDatalist } from './accessibility-datalist.js';

> 

Author: ```suggestion

PatternFly provides accessible components, but **we can't guarantee that your products will be accessible**. To consider your product accessible, you must take care to ensure that your product's components are accessible, which may include additional development work on your end.

```

## Prioritizing accessibility during development

As you develop your product, it can be helpful to keep accessibility top of mind, by asking yourself a few questions: 

**1. Can all users discover and perceive elements?**

> If you can see or click on it, then it should be discoverable and perceivable by all users, including those using a keyboard or screen reader.

Reviewer 2: Do you think it's clear that discoverable/perceivable means users can navigate to the element? When I wrote that second sentence, mostly I was trying to say it a different way to make sure that the audience understood what I meant. If you think it's clear, I'm totally good with this change, but just want to make sure. :)

Reviewer: Maybe adding in a ", and thus navigable" or something?

Only additional thing I have to add here is "...including those using a keyboard or other assistive technology such as a screen reader."

> If you can see or click on it, then it should be discoverable and perceivable by all users, including those using a keyboard or screen reader.

Author: ```suggestion
If you can see or click on it, then all users should be able to locate and navigate to it, including those using a keyboard or other assistive technology such as a screen reader.
```

**2. Can all users interact with elements?**

It should be easy for users to use an element once it's in focus. They should be able to interact with elements and initiate actions via keyboard (for example, pressing **Enter** to initiate a button action) and screen readers.

**3. Can all users understand elements?**

It should be clear what actions elements can perform. For example, buttons should have visible text that would be clear out of context of the page. If not, it should have an aria-label or accessible name.

> ## Extra steps for accessibility 

Reviewer 2: I have some concerns about this title, only because in the past I frequently faced the belief that using a PF component would automatically mean that a product's UI is 100% accessible. Calling it "Extra steps" makes me worried that people will skim over that section, assuming that was if they wanted to do extra credit so to speak, which most developers won't feel like they have time to do. I'd rather title this something that doesn't make it sound not necessary, you know? I'm terrible at naming so I worry to give different suggestions, but I can try if it'd be helpful. 😄

Reviewer: What about something like "Considerations to take as a developer"

> ## Extra steps for accessibility 

Author: ```suggestion
## Additional development considerations
```

> PatternFly provides accessible components, but **we can't guarantee that your products will be accessible**. There are items outside the scope of PatternFly, so you should keep these in mind to ensure accessibility: 

Author: ```suggestion
There are items outside the scope of PatternFly, so you should keep these in mind to ensure accessibility: 
```

<br />

<AccessibilityDatalist />

<br />

> ## Recommendations for UI elements

Reviewer: Would this section make sense as a sub-section of the previous one? Like:

```
## Prioritizing accessibility during development
## Extra steps for accessibility
### Recommendations for UI elements
#### Labeling with ARIA and so on...
```

It might also just be a case of  what should be placed in that table vs this "Recommendations..." section. I may also be feeling @jessiehuff comment above about whether this section would come across to a consumer as "It's recommended so it's optional" or something 😆

> ## Recommendations for UI elements

Author: ```suggestion
## Considerations for UI elements
```

The [WCAG 2.0 techniques](https://www.w3.org/TR/WCAG20-TECHS/Overview.html#contents) provide examples on how to meet accessibility guidelines. The following techniques are standard across PatternFly for specific patterns.

> ### Labeling with ARIA

Author: ```suggestion
#### Labeling with ARIA
```

For sighted users, the context and visual appearance of an element can help users understand its purpose. But this isn’t always the case. For example, the X, often used in the top-right corner of pop-ups to indicate the control for closing, might not be clear to those using assistive technology: 

<img src="./Popover.png" alt="Popover example of close button" />

ARIA supplements HTML so that common application interactions and features can be passed to assistive technologies when there is not a native mechanism. This is typically done by including an `aria-label` that provides descriptive context to assistive technologies. **Only use labels when they are necessary to make an element accessible.** ARIA is meant to add context when there *isn’t* visible text, not to override or reiterate what’s already there. Make sure that your labels are useful and not redundant.

Note that ARIA can only affect the semantics of an element; it has no effect on its behavior. For example, while you can make an element hidden to screen readers with `aria-hidden=”true”`, that does not hide the element from focus.

> ### Form fields

Author: ```suggestion
#### Form fields
```

Use [explicit linking](https://www.w3.org/TR/WCAG20-TECHS/H44.html) between `label` and form input elements (such as `input`, `textarea`, or `select`) when both are present. [This "H44" guidelines](//www.w3.org/TR/WCAG20-TECHS/H44.html) provides an accessible name to screen readers and makes the label clickable to increase the overall clickable area of the form element.

When a `label` element cannot accompany a form input element, provide one using using `aria-label` or `aria-labelledby`, as outline by the ["ARIA14" guidelines](//www.w3.org/TR/WCAG20-TECHS/ARIA14.html). 

In a single-field form, the submit button label can serve as the field label for sighted users, [following the "G167" guidelines](//www.w3.org/TR/WCAG20-TECHS/general.html#G167). For assistive devices, use `aria-labelledby`.

> ### Icons

Author: ```suggestion
#### Icons
```

To render icons as something that assistive devices can ignore, pass `aria-hidden='true'` to all icons, either to the icon element or a parent element of the icon. 

> #### **Decorative and semantic icons**

Author: ```suggestion
##### **Decorative and semantic icons**
```

There are two main categories of accessible icons:
1. **Decorative icons**, which can be removed without affecting the page's information.
1. **Semantic icons**, which provide information that isn't present on the page. For example, these can represent status, indicate an alert message type, or replace text as button labels. The meaning of a semantic icon must be provided in alternative ways to the user. Additionally, you should follow these guidelines:
    - Add a label for the icon in tooltip text that displays on hover and focus.
    - For interactive elements (like `<a>` and `<button>`), where an icon is used as the label instead of text, set a label as an `aria-label`. 

      Example: 

        ```html noLive
        <button class="..." aria-label="Close Dialog">
          <i class="..." aria-hidden="true"></i>
        </button>
        ```

> #### **Non-interactive icons**

Author: ```suggestion
##### **Non-interactive icons**
```

For **non-interactive icons**, include `.pf-screen-reader` text near the icon. Depending on the component, this text might not be a direct sibling to the icon element.  

Example: 

> In the alert component, the icon label text is adjacent to the message. This way, when `role="alert"` is added to `.pf-v5-c-alert__body` for dynamically displayed alerts, the type of message is announced along with the message text.

Author: Would the whole example be removed, then? I removed it in my last push to visualize, but can add it back and modify.

> In the alert component, the icon label text is adjacent to the message. This way, when `role="alert"` is added to `.pf-v5-c-alert__body` for dynamically displayed alerts, the type of message is announced along with the message text.

Reviewer: I think keeping the example could be useful, at least to show how using hidden text can be used.

> In the alert component, the icon label text is adjacent to the message. This way, when `role="alert"` is added to `.pf-v5-c-alert__body` for dynamically displayed alerts, the type of message is announced along with the message text.

Reviewer 2: We can definitely update this example to be more current. A lot of this content has been around for years, and I know that a few years back I did a good bit of work to try to get alerts to work across screen readers so I'm sure that things have changed since this content was originally created. So no objection to removing that part of the example. :)

**Note**: Alt text does not necessarily describe the visual characteristics of the image itself. It just needs to convey the same meaning as the image. 

The exception to this practice is when images are primarily for presentation purposes and are not essential pieces of content. To signify that an image should be skipped by a screen reader, set the value of the alt attribute to an empty string: `alt=””`.

### Trapping focus

The recommended interaction pattern for components like the modal or popover is to trap focus within the component's modal element when it becomes visible.  

For keyboard-only users who use the tab key to navigate the interface, trapped focus means that focus cannot be shifted outside of the modal when using the tab key. Instead, when focus leaves the last focusable item, it should move to the first focusable item of the modal. 

For screen reader users, the other contents on the page should be hidden from the screen reader.

> Based on the [screen reader / browser combinations](https://www.patternfly.org/v4/accessibility/testing-your-accessibility) we use for testing, we recommend applying `aria-hidden="true"` to the parent wrapping element of the page contents. The modal element of the component must not be a descendent of this element, and should be included as a sibling to this element.

Author: ```suggestion
Based on the [screen reader / browser combinations](https://www.patternfly.org/accessibility/testing-your-accessibility) we use for testing, we recommend applying `aria-hidden="true"` to the parent wrapping element of the page contents. The modal element of the component must not be a descendent of this element, and should be included as a sibling to this element.
```

        <p>Content</p>
    </div>
</div>
```

## Screen reader considerations

> Just as front-end developers use their browser to see how their changes look, you should use a screen reader to see how your accessibility looks (we use Voice Over). You can use any screen reader that is available in your operating system. To test PatternFly, we target the following screen readers:

Reviewer 2: The last line seems odd, especially because we don't mention any screen readers afterwards.

Reviewer 2: Also, I'm wondering if we could update the one line about screen readers in your operating system? VO is found in the mac operating system, but common ones like NVDA and JAWS aren't, they're downloads for PC. Maybe we could say something like, "You can use a screen reader that is available in your operating system, like VO for Mac, or you could download an open source screen reader like [NVDA](https://www.nvaccess.org/download/) if you have a PC." Feel free to wordsmith that to sound better though. 😂

> Just as front-end developers use their browser to see how their changes look, you should use a screen reader to see how your accessibility looks (we use Voice Over). You can use any screen reader that is available in your operating system. To test PatternFly, we target the following screen readers:

Author: ```suggestion
Just as front-end developers use their browser to see how their changes look, you should use a screen reader to see how your accessibility looks (we use VoiceOver). You can use any screen reader that is available in your operating system. To test PatternFly, we target the following screen readers:
```
